Compilation error on ternary operator [$10]

```zep
namespace Project;

class TestClass
{

	public static function doSomething()
	{
		var var2, var3;

		let var2 = ["asd","qwe"];

		let var3 = explode("delimiter", isset(var2[1]) ? var2[1] : "");
	}

}
```

In compile-errors.log:

```
.../testclass.zep.c:43:2: error: member reference base type 'zval *' (aka 'struct _zval_struct *') is not a structure or union
        ZEPHIR_SINIT_VAR(_1);
        ^~~~~~~~~~~~~~~~~~~~
./kernel/memory.h:79:2: note: expanded from macro 'ZEPHIR_SINIT_VAR'
        INIT_PZVAL(&z); \
        ^~~~~~~~~~~~~~
./Zend/zend.h:732:5: note: expanded from macro 'INIT_PZVAL'
        (z)->refcount__gc = 1;  \
        ~~~^ ~~~~~~~~~~~~
./testclass.zep.c:43:2: error: member reference base type 'zval *' (aka 'struct _zval_struct *') is not a structure or union
        ZEPHIR_SINIT_VAR(_1);
        ^~~~~~~~~~~~~~~~~~~~
./kernel/memory.h:79:2: note: expanded from macro 'ZEPHIR_SINIT_VAR'
        INIT_PZVAL(&z); \
        ^~~~~~~~~~~~~~
./Zend/zend.h:733:5: note: expanded from macro 'INIT_PZVAL'
        (z)->is_ref__gc = 0;
        ~~~^ ~~~~~~~~~~
./testclass.zep.c:43:2: error: member reference type 'zval *' (aka 'struct _zval_struct *') is a pointer; maybe you meant to use '->'?
        ZEPHIR_SINIT_VAR(_1);
        ^~~~~~~~~~~~~~~~~~~~
./kernel/memory.h:80:2: note: expanded from macro 'ZEPHIR_SINIT_VAR'
        ZVAL_NULL(&z);
        ^~~~~~~~~~~~~
.../Zend/zend_API.h:555:3: note: expanded from macro 'ZVAL_NULL'
                Z_TYPE_P(z) = IS_NULL;          \
                ^~~~~~~~~~~
.../Zend/zend_operators.h:461:26: note: expanded from macro 'Z_TYPE_P'
#define Z_TYPE_P(zval_p)        Z_TYPE(*zval_p)
                                ^~~~~~~~~~~~~~~
.../Zend/zend_operators.h:460:29: note: expanded from macro 'Z_TYPE'
#define Z_TYPE(zval)            (zval).type
                                ~~~~~~^
```

In testclass.zep.c:

```cpp

PHP_METHOD(Project_TestClass, doSomething) {

	zval *var2, *var3, *_0 = NULL, *_1 = NULL;

	ZEPHIR_MM_GROW();

	ZEPHIR_INIT_VAR(var2);
	array_init_size(var2, 3);
	ZEPHIR_INIT_VAR(_0);
	ZVAL_STRING(_0, "asd", 1);
	zephir_array_fast_append(var2, _0);
	ZEPHIR_INIT_NVAR(_0);
	ZVAL_STRING(_0, "qwe", 1);
	zephir_array_fast_append(var2, _0);
	ZEPHIR_INIT_VAR(var3);
	ZEPHIR_SINIT_VAR(_1); // 43
	if (zephir_array_isset_long(var2, 1)) {
		ZEPHIR_OBS_NVAR(_1);
		zephir_array_fetch_long(&_1, var2, 1, PH_NOISY, "project/TestClass.zep", 12 TSRMLS_CC);
	} else {
		ZEPHIR_INIT_NVAR(_1);
		ZVAL_STRING(_1, "", 1);
	}
	zephir_fast_explode_str(var3, SL("delimiter"), _1, LONG_MAX TSRMLS_CC);
	ZEPHIR_MM_RESTORE();

}

``` 

Zephir version 0.5.9a (ee96c7e)
PHP 5.4.35

<bountysource-plugin>
---
There is a **[$10 open bounty](https://www.bountysource.com/issues/6729099-compilation-error-on-ternary-operator?utm_campaign=plugin&utm_content=tracker%2F280146&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F280146&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/Library/Operators/Other/TernaryOperator.php
+++ b/Library/Operators/Other/TernaryOperator.php
@@ -44,9 +44,12 @@ class TernaryOperator extends BaseOperator
     public function compile($expression, CompilationContext $compilationContext)
     {
         /**
-         * This variable is used to check if the compound and expression is evaluated as true or false
+         * This variable is used to check if the compound and expression is evaluated as true or false:
+         * Ensure that newly allocated variables are local-only (setReadOnly)
          */
+        $this->setReadOnly(false);
         $returnVariable = $this->getExpected($compilationContext, $expression, false);
+        /* Make sure that passed variables (passed symbol variables) are promoted */
         $returnVariable->setLocalOnly(false);
 
         if ($returnVariable->getType() != 'variable' || $returnVariable->getName() == 'return_value') {

--- a/test/ternary.zep
+++ b/test/ternary.zep
@@ -32,6 +32,20 @@ class Ternary
 		//return gettype(typeof a == "resource" ? "unknown": false);
 	}
 
+	/**
+	 * @link https://github.com/phalcon/zephir/issues/665
+	 */
+	public function testTernaryWithPromotedTemporaryVariable()
+	{
+		var var2, var3;
+
+		let var2 = ["_b_","_c_"];
+
+		let var3 = explode("_", isset(var2[1]) ? var2[1] : "");
+
+		return var3;
+	}
+
     /**
      * @link https://github.com/phalcon/zephir/issues/297
      */

--- a/unit-tests/Extension/TernaryTest.php
+++ b/unit-tests/Extension/TernaryTest.php
@@ -28,6 +28,7 @@ class TernaryTest extends \PHPUnit_Framework_TestCase
         $this->assertEquals('foo', $t->testTernary2(true));
         $this->assertEquals('bar', $t->testTernary2(false));
         $this->assertEquals(3, $t->testTernaryAfterLetVariable());
+        $this->assertEquals(array('', 'c', ''), $t->testTernaryWithPromotedTemporaryVariable());
     }
 
     /*public function testComplex()

